storage/txcache/txByHashMap.go

```diff
@@ -58,15 +58,15 @@ func (txMap *txByHashMap) getTx(txHash string) (*WrappedTransaction, bool) {
 
 // RemoveTxsBulk removes transactions, in bulk
 func (txMap *txByHashMap) RemoveTxsBulk(txHashes [][]byte) uint32 {
-	oldCount := uint32(txMap.counter.Get())
+	numRemoved := uint32(0)
 
 	for _, txHash := range txHashes {
-		txMap.removeTx(string(txHash))
+		_, removed := txMap.removeTx(string(txHash))
```

Comment:
the removeTx implementation seems a bit weak as for example if we concurrently try to remove the same txhash we might end up with a wrong count and wrong size (removal succeeds only once but we decrement the counter and the size twice)

Reply:
Good point!

`tx, ok := txMap.getTx(txHash)` could return `ok == true` for both the concurring routines and both will decrement, which, indeed, is not correct.

The issue can appear if transactions are removed during eviction (on high load) and also removed (at the same time) on the "commit / finalize block" event

In `monitoring.go`, in function `diagnose()` we have:

```
numTxsEstimate := int(cache.CountTx()) // this comes from the possibly wrong counter
numTxsInChunks := cache.txByHash.backingMap.Count()
```

If the estimate does not equal `numTxsInChunks`, a `log.Debug()` is emitted and we should see the difference. This `diagnose()` is unfortunately (currently) called only upon transaction selection (on **my turn**).

A possible solution to this issue: in `txMap.backingMap.Remove`, when the map chunk is locked (under mutex), do the "item exists" check and call a `onItemRemovedCallback()` to decrement the atomic counters. Would this be fine?

Another solution would be to apply corrections from time to time - that would work fine for the counter, but the corrections for the _size counter_ would be quite inefficient.

I've added a TODO for this correction (it's an old condition).

Reply:
I was thinking the same, add the check in the Remove under lock.